### ISSUE TITLE
security configuration issue with opensearch multi-node cluster setup (in k8s env)

### ISSUE THREAD

**Issue by nickytd** [opensearch-project/OpenSearch#987]
Hi, 

I am spending some time these days to prepare a helm chart for setting up a complete (scalable) opensearch logging stack in a kubernetes env. The [AWS Opendistro for Elasticsearch](https://github.com/nickytd/kubernetes-logging-helm/tree/master) setup works ok and now the effort is dedicated to get the helm chart supporting also [opensearch](https://github.com/nickytd/kubernetes-logging-helm/tree/with-opensearch)

Here is the current state for the opensearch case:
When a single node setup is used then the cluster is initialized properly. Single node type can be successfully scaled up and everything is functioning as expected.

However the chart also supports separating node types (coordination, data and master) and allows scaling up those independently.  In this case setting up the security configuration although applied successfully doesn't bring the cluster in the expected functional state. 

The same config works just fine when all nodes share same (all) types. I'd like to ask for a help to get the multi type setup also in a correct functional state. Please drop me an email and we can organise a remote session where I can go into details of the setup and show where I see the problems. Meanwhile I will continue trying to find a solution. 
Will follow up if successful. 

Thanks
Niki Dokovski

**Comment by dblock** [comment]
@nickytd Thanks for reporting this. It would be great if you could try to post a way to reproduce this here, rather than trying to get someone to do remote sessions. It's more work, but this way we can build something repeatable once we identify the problem. Let's do remote as a last resort?

**Comment by nickytd** [comment]
Yes sure. The helm charts has [examples](https://github.com/nickytd/kubernetes-logging-helm/tree/with-opensearch/examples) folder which contains configurations for the various setup modes.
So in short to test the different modes.
1. get a minikube and start it with 2 nodes (minikube start --nodes=2)
2. checkout the helm chart and switch to -with-opensearch branch
3. Make sure current kubeconfig targets the minikube cluster and run the setup.sh in the examples folder. 
This brings the cluster setup with a single opensearch node, fluent bit and opensearch-dashboard. The scenario works as expected. You can authenticate in the opensearch-dashboards, check dashboards, switch tenants etc.

To simulate the multi node setup, get a clean minikube cluster, (with 2 nodes) and execute this time the setup.sh --scaled from the examples folder. This will bring up (2 coordination, 2 master 2 data nodes, 2 kafka brokers , 2 zookeepers nodes, fluendbit and fluentd) so it may take some time. This time however the operational state of cluster is not correct.

The main branch of the repo contains the current related version of the helm chart based on AWS OpenDistro for Elasticsearch. And the goal here is to release it with opensearch 1.0 GA. 

Meanwhile I also continue looking into the issue and will update in case of a progress. Any other feedback is of course more than welcome
Thanks
Niki Dokovski

**Comment by nickytd** [comment]
The issue goes down to a problem with the index mappings.
There are [predefined templates](https://github.com/nickytd/kubernetes-logging-helm/tree/with-opensearch/charts/index-templates) which are uploaded to the cluster during the setup phase. 
However when fluentd starts to put (_bulk) logs to the opensearch the data node rejects those with 
[2021-07-23T07:40:52,547][INFO ][o.o.a.b.TransportShardBulkAction] [efk-data-0] [containers-2021.07.23][0] mapping update rejected by primary
java.lang.IllegalArgumentException: mapper [kubernetes.pod_name] cannot be changed from type [keyword] to [text]
	at org.opensearch.index.mapper.ParametrizedFieldMapper.merge(ParametrizedFieldMapper.java:103) ~[opensearch-1.0.0.jar:1.0.0]

And when looking into the index settings the template mapping is indeed not there. 
{
    "systemd-2021.07.23": {
        "settings": {
            "index": {
                "codec": "best_compression",
                "mapping": {
                    "total_fields": {
                        "limit": "2000"
                    }
                },


The index templates are uploaded and present in the cluster (at least from the maser node logs) long before any indexing starts and at least the master node acknowledges this and we can see this in the logs

[2021-07-23T07:40:01,370][INFO ][o.o.c.m.MetadataCreateIndexService] [efk-master-0] [containers-2021.07.23] creating index, cause [auto(bulk api)], templates [containers], shards [1]/[0] 

In summary:
Master Node: creates index with a template
Data Node: rejects indexing as it doesn't see the index template

Thanks
Niki Dokovski
PS / this very same configuration is fully functional when all nodes share same types and there are no separation on coordination, master and data nodes /
